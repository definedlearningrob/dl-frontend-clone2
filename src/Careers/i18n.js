import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import { merge } from 'lodash-es';

import translation from '@dc/locales/en-us/translation';

import sharedTranslation from '@shared/locales/en-us/translation';

i18n.use(initReactI18next).init({
  fallbackLng: 'en-us',
  debug: import.meta.env.MODE === 'development',
  lng: 'en-us',
  resources: {
    'en-US': {
      translation: merge(translation, sharedTranslation),
    },
  },
  interpolation: { escapeValue: false },
});

export default i18n;
