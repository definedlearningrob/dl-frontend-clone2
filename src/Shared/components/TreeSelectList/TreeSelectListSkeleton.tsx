import SkeletonRectangle from '../Skeleton/Rectangle/Rectangle';

export const TreeSelectListSkeleton = () => (
  <div className='h-[280px] py-xs'>
    <SkeletonRectangle className='my-xs' height='base' radius='sm' />
    <SkeletonRectangle className='my-xs' height='base' radius='sm' />
    <SkeletonRectangle className='my-xs' height='base' radius='sm' />
    <SkeletonRectangle className='my-xs' height='base' radius='sm' />
  </div>
);
