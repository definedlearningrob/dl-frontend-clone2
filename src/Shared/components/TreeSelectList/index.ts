export * from './TreeSelectList';
export type { TreeSelectOption } from './TreeSelectListItem';
