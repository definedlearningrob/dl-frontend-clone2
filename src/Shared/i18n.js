import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

import sharedTranslation from '@shared/locales/en-us/translation';

i18n.use(initReactI18next).init({
  fallbackLng: 'en-us',
  // eslint-disable-next-line no-undef
  debug: import.meta.env.MODE === 'development',
  lng: 'en-us',
  resources: {
    'en-US': {
      translation: sharedTranslation,
    },
  },
  interpolation: { escapeValue: false },
});

export default i18n;
