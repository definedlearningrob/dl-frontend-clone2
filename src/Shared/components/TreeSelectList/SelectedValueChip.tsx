import { useTranslation } from 'react-i18next';

import SharedAvatar from '@shared/components/Avatar/Avatar';
import DeprecatedIconButton from '@shared/components/DeprecatedIconButton/DeprecatedIconButton';
import { ReactComponent as RemoveIcon } from '@shared/svg/close.svg';
import { Tooltip } from '@shared/components/Tooltip';

import { TreeSelectOption } from './TreeSelectListItem';

type Props = {
  option: TreeSelectOption;
  onRemove: (value: string) => void;
};

export const SelectedValueChip = ({ option, onRemove }: Props) => {
  const { t } = useTranslation();
  const { label, value, withAvatar } = option;

  const handleRemove = () => {
    onRemove(value);
  };

  return (
    <div
      className='flex gap-xxs items-center p-xxs rounded-xs bg-neutral-200'
      data-testid='selected-value-chip'>
      <div className='flex items-center gap-xs'>
        {withAvatar && (
          <div className='rounded-full outline outline-1 outline-neutral-300'>
            <SharedAvatar label={label} size='24' theme='light' />
          </div>
        )}
        <span className='whitespace-nowrap text-xs text-font-secondary'>{label}</span>
      </div>
      <Tooltip delayDuration={500} message={t('common.actions.remove')}>
        <DeprecatedIconButton
          aria-label={t('components.treeSelectList.removeOption', { option: label })}
          className='!p-0'
          icon={<RemoveIcon />}
          onClick={handleRemove}
        />
      </Tooltip>
    </div>
  );
};
