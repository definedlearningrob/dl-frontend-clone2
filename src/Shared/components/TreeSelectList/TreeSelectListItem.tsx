import { isEmpty, uniq, xor } from 'lodash-es';
import { useToggle } from 'react-use';
import { MouseEvent, useMemo } from 'react';
import { useTranslation } from 'react-i18next';

import { ReactComponent as ChevronUpIcon } from '@shared/svg/chevron_up.svg';
import { ReactComponent as ChevronDownIcon } from '@shared/svg/chevron_down.svg';
import SharedCheckbox from '@shared/components/Checkbox/Checkbox';
import { cx } from '@shared/utils/cx';
import SharedAvatar from '@shared/components/Avatar/Avatar';
import { Tooltip } from '@shared/components/Tooltip';
import { IconButton } from '@shared/components/IconButton/IconButton';

import { flattenOptions } from './helpers';

export interface TreeSelectOption {
  label: string;
  value: string;
  withAvatar?: boolean;
  children: this[];
}

type Props = {
  option: TreeSelectOption;
  selectedValues: string[];
  onChange: (values: string[]) => void;
  nestingLevel?: number;
};

export const TreeSelectListItem = ({
  option,
  selectedValues,
  onChange,
  nestingLevel = 0,
}: Props) => {
  const { t } = useTranslation();
  const [isExpanded, toggleIsExpanded] = useToggle(false);
  const { value, label, children, withAvatar } = option;

  const ExpandIcon = isExpanded ? ChevronUpIcon : ChevronDownIcon;
  const isRootOption = nestingLevel === 0;
  const hasChildren = !isEmpty(children);

  const flattenedChildren = useMemo(() => flattenOptions(children), [children]);

  const hasSelectedChildren = flattenedChildren.some((option) =>
    selectedValues.includes(option.value)
  );

  const getIsSelected = () => {
    if (!hasChildren) {
      return selectedValues.includes(value);
    }

    return flattenedChildren.every((option) => selectedValues.includes(option.value));
  };

  const isSelected = getIsSelected();

  const handleChange = () => {
    if (!hasChildren) {
      onChange(xor(selectedValues, [option.value]));

      return;
    }

    const childrenValues = flattenedChildren.map(({ value }) => value);
    const newValue = isSelected
      ? selectedValues.filter((option) => !childrenValues.includes(option))
      : uniq([...selectedValues, ...childrenValues]);

    onChange(newValue);
  };

  const handleToggleExpanded = (event: MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    toggleIsExpanded();
  };

  const optionClasses = cx(
    'flex items-center gap-xs py-xs pr-xs w-full border-b border-neutral-300 cursor-pointer',
    'focus-visible:bg-primary-200 focus-within:bg-primary-200',
    { 'bg-primary-200': isSelected }
  );

  return (
    <>
      <li
        aria-label={label}
        className={optionClasses}
        style={{ paddingLeft: isRootOption ? 8 : nestingLevel * 64 }}
        onClick={handleChange}>
        {hasChildren && (
          <Tooltip
            delayDuration={500}
            message={isExpanded ? t('common.actions.fold') : t('common.actions.expand')}>
            <IconButton
              Icon={ExpandIcon}
              aria-label={isExpanded ? t('common.actions.fold') : t('common.actions.expand')}
              onClick={handleToggleExpanded}
            />
          </Tooltip>
        )}
        <div className='flex items-center'>
          <SharedCheckbox
            checked={isSelected}
            indeterminate={!isSelected && hasSelectedChildren}
            onChange={handleChange}
          />
          <div className='flex items-center gap-xs'>
            {withAvatar && (
              <div className='rounded-full outline outline-1 outline-neutral-300 ml-xs'>
                <SharedAvatar label={label} size='24' theme='light' />
              </div>
            )}
            {label}
          </div>
        </div>
      </li>
      {isExpanded &&
        children.map((option) => (
          <TreeSelectListItem
            key={option.value}
            nestingLevel={nestingLevel + 1}
            option={option}
            selectedValues={selectedValues}
            onChange={onChange}
          />
        ))}
    </>
  );
};
