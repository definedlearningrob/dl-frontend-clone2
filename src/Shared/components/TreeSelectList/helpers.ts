import { flatMapDeep } from 'lodash-es';

import { TreeSelectOption } from './TreeSelectListItem';

export const flattenOptions = (options: TreeSelectOption[]): TreeSelectOption[] =>
  flatMapDeep(options, (option) => [{ ...option, children: [] }, flattenOptions(option.children)]);
