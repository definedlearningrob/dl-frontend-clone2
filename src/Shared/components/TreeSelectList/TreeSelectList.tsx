import { isEmpty } from 'lodash-es';
import { ChangeEvent, useMemo, useState } from 'react';

import { cx } from '@shared/utils/cx';
import { IconContainer } from '@shared/components/IconContainer/IconContainer';
import { ReactComponent as SearchIcon } from '@shared/svg/search.svg';

import { TreeSelectListItem, TreeSelectOption } from './TreeSelectListItem';
import { SelectedValueChip } from './SelectedValueChip';
import { flattenOptions } from './helpers';
import { TreeSelectListSkeleton } from './TreeSelectListSkeleton';

type Props = {
  options: TreeSelectOption[];
  placeholder: string;
  label: string;
  isLoading?: boolean;
  isRequired?: boolean;
  value: string[];
  onChange: (newValue: string[]) => void;
};

export const TreeSelectList = ({
  options,
  placeholder,
  label,
  isLoading,
  isRequired,
  value,
  onChange,
}: Props) => {
  const [searchValue, setSearchValue] = useState('');

  const hasSelectedValues = !isEmpty(value);

  const normalizedOptions = useMemo(
    () =>
      flattenOptions(options).reduce((acc, option) => {
        acc[option.value] = option;

        return acc;
      }, {} as { [key: string]: TreeSelectOption }),
    [options]
  );

  const selectedOptions = value.map((item) => normalizedOptions[item]);

  const handleRemoveValue = (valueToRemove: string) => {
    const filteredValues = value.filter((value) => value !== valueToRemove);
    onChange(filteredValues);
  };

  const handleChangeSearchValue = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchValue(event.target.value);
  };

  const getFilterOptions = (options: TreeSelectOption[]): TreeSelectOption[] =>
    options.flatMap((option) => {
      if (option.label.toLowerCase().includes(searchValue.toLowerCase())) {
        return [option];
      }

      const children = getFilterOptions(option.children);

      if (children.length > 0) {
        return [{ ...option, children }];
      }

      return [];
    });

  return (
    <>
      <div className='mb-sm'>
        <label className='flex gap-xxs text-xs leading-lg mb-xs' htmlFor='tree-select-input'>
          {isRequired && <span className='text-danger-600'>*</span>}
          {label}
        </label>
        <div
          className={cx(
            'flex items-center flex-wrap gap-xs bg-white transition-colors border rounded-sm max-h-[122px] scrollbar',
            'border-neutral-300 hover:border-neutral-400 focus-within:!border-primary-500',
            {
              'px-xs py-xxs': hasSelectedValues,
              'p-xs': !hasSelectedValues,
            }
          )}>
          {selectedOptions.map((option) => (
            <SelectedValueChip key={option!.value} option={option!} onRemove={handleRemoveValue} />
          ))}
          {!hasSelectedValues && (
            <IconContainer Icon={SearchIcon} className='text-neutral-400' paddingSize='none' />
          )}
          <input
            className='flex-1 min-w-[25%] outline-none text-xs placeholder:text-font-secondary'
            id='tree-select-input'
            placeholder={hasSelectedValues ? '' : placeholder}
            value={searchValue}
            onChange={handleChangeSearchValue}
          />
        </div>
      </div>
      {isLoading && <TreeSelectListSkeleton />}
      {!isLoading && (
        <ul className='h-[280px] scrollbar'>
          {getFilterOptions(options).map((option) => (
            <TreeSelectListItem
              key={option.value}
              option={option}
              selectedValues={value}
              onChange={onChange}
            />
          ))}
        </ul>
      )}
    </>
  );
};
