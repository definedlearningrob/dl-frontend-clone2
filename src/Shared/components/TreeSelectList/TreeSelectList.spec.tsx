import { render, screen, within } from '@testing-library/react';
import { noop } from 'lodash-es';
import userEvent from '@testing-library/user-event';
import { I18nextProvider } from 'react-i18next';

import i18n from '@shared/i18n';

import { TreeSelectList } from './TreeSelectList';

const options = [
  {
    label: 'Calculus',
    value: '1',
    children: [
      {
        label: 'John Doe',
        value: '11',
        children: [],
        withAvatar: true,
      },
      {
        label: 'Jennifer Fox',
        value: '12',
        children: [],
        withAvatar: true,
      },
    ],
  },
  {
    label: 'Statistics',
    value: '2',
    children: [
      {
        label: 'Amy Smith',
        value: '21',
        children: [],
        withAvatar: true,
      },
    ],
  },
];

const renderComponent = (value: string[], onChange = noop) =>
  render(
    <I18nextProvider i18n={i18n}>
      <TreeSelectList
        label='Label'
        options={options}
        placeholder='Select options'
        value={value}
        onChange={onChange}
      />
    </I18nextProvider>
  );

describe('TreeSelectList', () => {
  it('renders correctly with selected values', () => {
    const { container } = renderComponent(['12', '21']);

    const selectedValues = screen.getAllByTestId('selected-value-chip');
    expect(selectedValues).toHaveLength(2);

    expect(selectedValues[0]).toHaveTextContent('Jennifer Fox');
    expect(selectedValues[1]).toHaveTextContent('Amy Smith');

    expect(container).toMatchSnapshot();
  });

  it('sets correct value when selecting parent option', () => {
    const onChange = jest.fn();
    renderComponent([], onChange);

    const parentOption = screen.getByRole('listitem', { name: 'Calculus' });
    userEvent.click(parentOption);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(['11', '12']);
  });

  it('sets correct value when deselecting parent option', () => {
    const onChange = jest.fn();
    renderComponent(['11', '12', '21'], onChange);

    const parentOption = screen.getByRole('listitem', { name: 'Calculus' });
    userEvent.click(parentOption);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(['21']);
  });

  it('toggles children option value', () => {
    const onChange = jest.fn();
    renderComponent(['11'], onChange);

    const parentOption = screen.getByRole('listitem', { name: 'Calculus' });
    const expandButton = within(parentOption).getByRole('button', { name: 'Expand' });
    userEvent.click(expandButton);

    const selectedOption = screen.getByRole('listitem', { name: 'John Doe' });
    const notSelectedOption = screen.getByRole('listitem', { name: 'Jennifer Fox' });

    userEvent.click(selectedOption);
    expect(onChange).toHaveBeenCalledWith([]);

    userEvent.click(notSelectedOption);
    expect(onChange).toHaveBeenCalledWith(['11', '12']);
  });

  it('removes value when clicking remove button on a chip', () => {
    const onChange = jest.fn();
    renderComponent(['11'], onChange);

    const removeButton = screen.getByRole('button', { name: 'Remove John Doe' });
    userEvent.click(removeButton);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith([]);
  });

  it('filters options correctly', () => {
    renderComponent([]);

    const searchInput = screen.getByRole('textbox', { name: 'Label' });
    userEvent.paste(searchInput, 'John');

    const options = screen.getAllByRole('listitem');
    expect(options).toHaveLength(1);
    expect(options[0]).toHaveTextContent('Calculus');

    const [parentOption] = options;
    const expandButton = within(parentOption).getByRole('button', { name: 'Expand' });
    userEvent.click(expandButton);

    const expandedOptions = screen.getAllByRole('listitem');
    expect(expandedOptions).toHaveLength(2);
    expect(expandedOptions[0]).toHaveTextContent('Calculus');
    expect(expandedOptions[1]).toHaveTextContent('John Doe');
  });
});
