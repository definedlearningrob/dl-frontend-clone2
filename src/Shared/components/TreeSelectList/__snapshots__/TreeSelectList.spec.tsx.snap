// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeSelectList renders correctly with selected values 1`] = `
<div>
  <div
    class="mb-sm"
  >
    <label
      class="flex gap-xxs text-xs leading-lg mb-xs"
      for="tree-select-input"
    >
      Label
    </label>
    <div
      class="flex items-center flex-wrap gap-xs bg-white transition-colors border rounded-sm max-h-[122px] scrollbar border-neutral-300 hover:border-neutral-400 focus-within:!border-primary-500 px-xs py-xxs"
    >
      <div
        class="flex gap-xxs items-center p-xxs rounded-xs bg-neutral-200"
        data-testid="selected-value-chip"
      >
        <div
          class="flex items-center gap-xs"
        >
          <div
            class="rounded-full outline outline-1 outline-neutral-300"
          >
            <div
              class="avatar"
              data-testid="avatar-image"
            >
              <div
                class="avatar__image sb-avatar"
                style="display: inline-block; vertical-align: middle; width: 24px; height: 24px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif;"
              >
                <div
                  class="avatar__image sb-avatar__text"
                  style="width: 24px; height: 24px; line-height: initial; text-align: center; color: rgb(106, 110, 124); background: rgb(252, 252, 252); border-radius: 100%;"
                  title="Jennifer Fox"
                >
                  <div
                    style="display: table; table-layout: fixed; width: 100%; height: 100%;"
                  >
                    <span
                      style="display: table-cell; vertical-align: middle; font-size: 100%; white-space: nowrap;"
                    >
                      <span />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <span
            class="whitespace-nowrap text-xs text-font-secondary"
          >
            Jennifer Fox
          </span>
        </div>
        <span
          data-state="closed"
        >
          <button
            aria-label="Remove Jennifer Fox"
            class="icon-button !p-0"
            data-testid="icon-button"
            type="button"
          >
            <div
              class="icon icon-button__icon"
              data-testid="icon"
            >
              <div />
            </div>
          </button>
        </span>
      </div>
      <div
        class="flex gap-xxs items-center p-xxs rounded-xs bg-neutral-200"
        data-testid="selected-value-chip"
      >
        <div
          class="flex items-center gap-xs"
        >
          <div
            class="rounded-full outline outline-1 outline-neutral-300"
          >
            <div
              class="avatar"
              data-testid="avatar-image"
            >
              <div
                class="avatar__image sb-avatar"
                style="display: inline-block; vertical-align: middle; width: 24px; height: 24px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif;"
              >
                <div
                  class="avatar__image sb-avatar__text"
                  style="width: 24px; height: 24px; line-height: initial; text-align: center; color: rgb(106, 110, 124); background: rgb(252, 252, 252); border-radius: 100%;"
                  title="Amy Smith"
                >
                  <div
                    style="display: table; table-layout: fixed; width: 100%; height: 100%;"
                  >
                    <span
                      style="display: table-cell; vertical-align: middle; font-size: 100%; white-space: nowrap;"
                    >
                      <span />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <span
            class="whitespace-nowrap text-xs text-font-secondary"
          >
            Amy Smith
          </span>
        </div>
        <span
          data-state="closed"
        >
          <button
            aria-label="Remove Amy Smith"
            class="icon-button !p-0"
            data-testid="icon-button"
            type="button"
          >
            <div
              class="icon icon-button__icon"
              data-testid="icon"
            >
              <div />
            </div>
          </button>
        </span>
      </div>
      <input
        class="flex-1 min-w-[25%] outline-none text-xs placeholder:text-font-secondary"
        id="tree-select-input"
        placeholder=""
        value=""
      />
    </div>
  </div>
  <ul
    class="h-[280px] scrollbar"
  >
    <li
      aria-label="Calculus"
      class="flex items-center gap-xs py-xs pr-xs w-full border-b border-neutral-300 cursor-pointer focus-visible:bg-primary-200 focus-within:bg-primary-200"
      style="padding-left: 8px;"
    >
      <span
        data-state="closed"
      >
        <button
          aria-label="Expand"
          class="flex items-center justify-center transition-colors rounded-sm border border-transparent cursor-pointer focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] text-primary-500 bg-white hover:bg-primary-200 h-base w-base "
          data-testid="icon-button"
          type="button"
        >
          <div
            class="flex"
          >
            <div
              class="flex align-center justify-center grow-0 shrink-0"
              data-testid="icon"
            >
              <div
                class="h-sm w-sm"
              />
            </div>
          </div>
        </button>
      </span>
      <div
        class="flex items-center"
      >
        <div
          class="checkbox"
        >
          <input
            class="checkbox__input"
            data-testid="checkbox"
            tabindex="0"
            type="checkbox"
          />
          <label
            class="checkbox__label"
          />
        </div>
        <div
          class="flex items-center gap-xs"
        >
          Calculus
        </div>
      </div>
    </li>
    <li
      aria-label="Statistics"
      class="flex items-center gap-xs py-xs pr-xs w-full border-b border-neutral-300 cursor-pointer focus-visible:bg-primary-200 focus-within:bg-primary-200 bg-primary-200"
      style="padding-left: 8px;"
    >
      <span
        data-state="closed"
      >
        <button
          aria-label="Expand"
          class="flex items-center justify-center transition-colors rounded-sm border border-transparent cursor-pointer focus-visible:outline focus-visible:outline-1 focus-visible:outline-offset-[3px] text-primary-500 bg-white hover:bg-primary-200 h-base w-base "
          data-testid="icon-button"
          type="button"
        >
          <div
            class="flex"
          >
            <div
              class="flex align-center justify-center grow-0 shrink-0"
              data-testid="icon"
            >
              <div
                class="h-sm w-sm"
              />
            </div>
          </div>
        </button>
      </span>
      <div
        class="flex items-center"
      >
        <div
          class="checkbox"
        >
          <input
            checked=""
            class="checkbox__input"
            data-testid="checkbox"
            tabindex="0"
            type="checkbox"
          />
          <label
            class="checkbox__label -checked"
          />
        </div>
        <div
          class="flex items-center gap-xs"
        >
          Statistics
        </div>
      </div>
    </li>
  </ul>
</div>
`;
